include/yarpgen/type.h

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright (c) 2019, Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <limits>
+#include <climits>
+
+#include "ir_value.h"
+
+namespace yarpgen {
+
+// Abstract class that serves as a common ancestor for all types.
+class Type {
+public:
+    enum class CVQualifier {
+        NONE,
+        CONST,
+        VOLAT,
+        CONST_VOLAT,
+    };
+
+    Type() : is_static(false), cv_qualifier(CVQualifier::NONE) {}
+    Type(bool _is_static, CVQualifier _cv_qual) : is_static(_is_static), cv_qualifier(_cv_qual) {}
+    virtual ~Type() = default;
+
+    virtual std::string getName() = 0;
+    virtual void dbgDump() = 0;
+
+    bool getIsStatic() { return is_static; }
+    void setIsStatic(bool _is_static) { is_static = _is_static; }
+    CVQualifier getCVQualifier() { return cv_qualifier; }
+    void setCVQualifier(CVQualifier _cv_qual) { cv_qualifier = _cv_qual; }
+
+private:
+    bool is_static;
+    CVQualifier cv_qualifier;
+    // We don't store name here because for compound types it might be tricky to get them.
+};
+
+// Base class for floating-point and integral types
+class ArithmeticType : public Type {
+public:
+    ArithmeticType() : Type() {}
+    ArithmeticType (bool _is_static, CVQualifier _cv_qual) : Type(_is_static, _cv_qual) {}
+    virtual std::string getLiteralSuffix() { return ""; };
+};
+
+class FPType : public ArithmeticType {
+    //TODO: it is a stub for now
+};
+
+class IntegralType : public ArithmeticType {
+public:
+    enum class IntTypeID {
+        BOOL,
+        // We don't have "char" as a type, because its signedness is implementation-defined
+        SCHAR,
+        UCHAR,
+        SHORT,
+        USHORT,
+        INT,
+        UINT,
+        LONG,
```

Comment:
Size of long is also platform dependent. It could make sense to have a separate class that define platform specific stuff. Though it's ok for now to just use these types.

Reply:
The size of the long will be set according to platform options inside the class.